File: backend/rust/services/game/src/domain/value_objects/bracket.rs
use mongodb::bson::{self, Bson};
use serde::{Deserialize, Serialize};
use shared::game::Actor;
use uuid::Uuid;

use crate::domain::value_objects::{LolTournamentSettings, participant::TeamParticipant};

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Match {
    pub game_series_id: Uuid,
    pub team1: Option<Actor>,
    pub team2: Option<Actor>,
    pub winner: Option<Actor>,
    pub round: usize,
    pub match_number: usize,
    pub next_match_id: Option<usize>, // ссылка на следующий матч (номер матча в следующем раунде)
    pub is_bye: bool,                 // является ли матч техническим (bye)
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Bracket {
    pub rounds: Vec<Vec<Match>>, // rounds[round_index][match_index]
    pub total_rounds: usize,
    pub total_matches: usize,
}

impl From<Bracket> for Bson {
    fn from(tp: Bracket) -> Self {
        bson::to_bson(&tp).unwrap()
    }
}

File: backend/python/game/domain/value_objects/bracket.py
from pydantic import BaseModel

from uuid import UUID

from domain.value_objects.actors import Actor


class Match(BaseModel):
    game_series_id: UUID
    team1: Actor | None
    team2: Actor | None
    winner: Actor | None
    round: int
    match_number: int
    next_match_id: int | None


class Bracket(BaseModel):
    rounds: list[list[Match]]

    def swap_teams(self, t1: Actor, t2: Actor) -> tuple[UUID, UUID]:
        m1, m2 = None, None
        for match in self.rounds[0]:
            if match.team1 == t1:
                m1 = (match, 0)
            elif match.team2 == t1:
                m1 = (match, 1)
            if match.team1 == t2:
                m2 = (match, 0)
            elif match.team2 == t2:
                m2 = (match, 1)
        if not m1 or not m2:
            raise Exception("Teams not found in bracket")
        match1, side1 = m1
        match2, side2 = m2
        if side1 == 0:
            match1.team1 = t2
        else:
            match1.team2 = t2
        if side2 == 0:
            match2.team1 = t1
        else:
            match2.team2 = t1
        return match1.game_series_id, match2.game_series_id
